File: go/rules.go
// Package quotagate mirrors the provider-style rate limits that model-serving
// APIs enforce: many limits per model applied together, across global/tenant/user
// scopes, with reserve-then-reconcile token accounting and standard back-pressure
// signals. It is the Go port of the Python reference implementation.
package quotagate

import (
	"encoding/json"
	"fmt"
	"strings"
)

// Scope is where a limit is enforced. A request is keyed differently per scope so
// the same rule isolates one global fleet, one tenant, or one user.
type Scope int

const (
	// Global enforces a single limit shared across the whole fleet.
	Global Scope = iota
	// Tenant enforces a limit per tenant.
	Tenant
	// User enforces a limit per user within a tenant.
	User
)

// String renders the scope as its lowercase wire name ("global"/"tenant"/"user").
func (s Scope) String() string {
	switch s {
	case Tenant:
		return "tenant"
	case User:
		return "user"
	default:
		return "global"
	}
}

// ParseScope maps a wire name to a Scope, returning an error for unknown names.
func ParseScope(s string) (Scope, error) {
	switch strings.ToLower(s) {
	case "global":
		return Global, nil
	case "tenant":
		return Tenant, nil
	case "user":
		return User, nil
	default:
		return Global, fmt.Errorf("unknown scope %q", s)
	}
}

// LimitRule is one provider-style limit. Any subset of the Max* dimensions may be
// set; a request must stay under every one that is.
type LimitRule struct {
	Model            string
	WindowSeconds    float64
	MaxTokens        *float64
	MaxRequests      *float64
	MaxCost          *float64
	MaxConcurrent    *int
	Scope            Scope
	BucketsPerWindow int
	Precise          bool
	Name             string
}

// RuleOption customises a LimitRule built with NewLimitRule.
type RuleOption func(*LimitRule)

// WithMaxTokens caps token volume over the window.
func WithMaxTokens(v float64) RuleOption { return func(r *LimitRule) { r.MaxTokens = &v } }

// WithMaxRequests caps request count over the window.
func WithMaxRequests(v float64) RuleOption { return func(r *LimitRule) { r.MaxRequests = &v } }

// WithMaxCost caps dollar spend over the window.
func WithMaxCost(v float64) RuleOption { return func(r *LimitRule) { r.MaxCost = &v } }

// WithMaxConcurrent caps the number of in-flight requests.
func WithMaxConcurrent(v int) RuleOption { return func(r *LimitRule) { r.MaxConcurrent = &v } }

// WithScope sets the scope the rule is enforced at.
func WithScope(s Scope) RuleOption { return func(r *LimitRule) { r.Scope = s } }

// WithBucketsPerWindow tunes the memory/accuracy trade-off of the bucketed window.
func WithBucketsPerWindow(n int) RuleOption { return func(r *LimitRule) { r.BucketsPerWindow = n } }

// WithPrecise switches the rule to the exact per-event log.
func WithPrecise(p bool) RuleOption { return func(r *LimitRule) { r.Precise = p } }

// WithName sets a human-readable label for the rule.
func WithName(name string) RuleOption { return func(r *LimitRule) { r.Name = name } }

// NewLimitRule builds a LimitRule with the standard defaults (Global scope, 60
// buckets per window) and applies the given options.
func NewLimitRule(model string, windowSeconds float64, opts ...RuleOption) LimitRule {
	r := LimitRule{Model: model, WindowSeconds: windowSeconds, Scope: Global, BucketsPerWindow: 60}
	for _, o := range opts {
		o(&r)
	}
	return r
}

// BucketSeconds returns the bucket width: 0 selects the exact per-event log;
// otherwise WindowSeconds/BucketsPerWindow.
func (r LimitRule) BucketSeconds() float64 {
	if r.Precise {
		return 0.0
	}
	buckets := r.BucketsPerWindow
	if buckets < 1 {
		buckets = 1
	}
	return r.WindowSeconds / float64(buckets)
}

// Label returns the rule's name, or a derived "model:scope:Ns" label.
func (r LimitRule) Label() string {
	if r.Name != "" {
		return r.Name
	}
	return fmt.Sprintf("%s:%s:%ds", r.Model, r.Scope, int64(r.WindowSeconds))
}

// HasUsageLimit reports whether the rule caps tokens, requests, or cost (as
// opposed to being a concurrency-only rule).
func (r LimitRule) HasUsageLimit() bool {
	return r.MaxTokens != nil || r.MaxRequests != nil || r.MaxCost != nil
}

// RulesFromJSON parses a rules document (see limits.sample.json) into LimitRules.
func RulesFromJSON(text string) ([]LimitRule, error) {
	var doc struct {
		Rules []struct {
			Model            string   `json:"model"`
			WindowSeconds    float64  `json:"window_seconds"`
			MaxTokens        *float64 `json:"max_tokens"`
			MaxRequests      *float64 `json:"max_requests"`
			MaxCost          *float64 `json:"max_cost"`
			MaxConcurrent    *int     `json:"max_concurrent"`
			Scope            *string  `json:"scope"`
			BucketsPerWindow *int     `json:"buckets_per_window"`
			Precise          *bool    `json:"precise"`
			Name             *string  `json:"name"`
		} `json:"rules"`
	}
	if err := json.Unmarshal([]byte(text), &doc); err != nil {
		return nil, err
	}
	out := make([]LimitRule, 0, len(doc.Rules))
	for _, row := range doc.Rules {
		scope := Global
		if row.Scope != nil {
			parsed, err := ParseScope(*row.Scope)
			if err != nil {
				return nil, err
			}
			scope = parsed
		}
		buckets := 60
		if row.BucketsPerWindow != nil {
			buckets = *row.BucketsPerWindow
		}
		precise := false
		if row.Precise != nil {
			precise = *row.Precise
		}
		name := ""
		if row.Name != nil {
			name = *row.Name
		}
		out = append(out, LimitRule{
			Model:            row.Model,
			WindowSeconds:    row.WindowSeconds,
			MaxTokens:        row.MaxTokens,
			MaxRequests:      row.MaxRequests,
			MaxCost:          row.MaxCost,
			MaxConcurrent:    row.MaxConcurrent,
			Scope:            scope,
			BucketsPerWindow: buckets,
			Precise:          precise,
			Name:             name,
		})
	}
	return out, nil
}

File: rust/src/rules.rs
//! Rule definitions and scopes: one provider-style limit per `LimitRule`, keyed
//! per scope so the same rule can isolate a fleet, a tenant, or a user.

use std::fmt;
use std::str::FromStr;

/// Where a limit is enforced. A request is keyed differently per scope so the
/// same rule isolates one global fleet, one tenant, or one user.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Scope {
    /// A single limit shared across the whole fleet.
    Global,
    /// A limit enforced per tenant.
    Tenant,
    /// A limit enforced per user within a tenant.
    User,
}

impl fmt::Display for Scope {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Scope::Global => "global",
            Scope::Tenant => "tenant",
            Scope::User => "user",
        };
        f.write_str(s)
    }
}

impl FromStr for Scope {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "global" => Ok(Scope::Global),
            "tenant" => Ok(Scope::Tenant),
            "user" => Ok(Scope::User),
            other => Err(format!("unknown scope \"{other}\"")),
        }
    }
}

/// One provider-style limit. Any subset of the `max_*` dimensions may be set; a
/// request must stay under every one that is.
#[derive(Debug, Clone)]
pub struct LimitRule {
    /// The model this rule applies to; `"*"` matches every model.
    pub model: String,
    /// The rolling window width in seconds.
    pub window_seconds: f64,
    /// Optional cap on token volume over the window.
    pub max_tokens: Option<f64>,
    /// Optional cap on request count over the window.
    pub max_requests: Option<f64>,
    /// Optional cap on dollar spend over the window.
    pub max_cost: Option<f64>,
    /// Optional cap on the number of in-flight requests.
    pub max_concurrent: Option<i64>,
    /// The scope the rule is enforced at.
    pub scope: Scope,
    /// The number of buckets the window is split into (memory/accuracy trade-off).
    pub buckets_per_window: i64,
    /// When true, keep an exact per-event log instead of buckets.
    pub precise: bool,
    /// Optional human-readable label.
    pub name: Option<String>,
}

impl LimitRule {
    /// Builds a rule with the standard defaults (Global scope, 60 buckets).
    pub fn new(model: &str, window_seconds: f64) -> Self {
        LimitRule {
            model: model.to_string(),
            window_seconds,
            max_tokens: None,
            max_requests: None,
            max_cost: None,
            max_concurrent: None,
            scope: Scope::Global,
            buckets_per_window: 60,
            precise: false,
            name: None,
        }
    }

    /// Sets the token cap.
    pub fn max_tokens(mut self, v: f64) -> Self {
        self.max_tokens = Some(v);
        self
    }

    /// Sets the request cap.
    pub fn max_requests(mut self, v: f64) -> Self {
        self.max_requests = Some(v);
        self
    }

    /// Sets the dollar-spend cap.
    pub fn max_cost(mut self, v: f64) -> Self {
        self.max_cost = Some(v);
        self
    }

    /// Sets the in-flight concurrency cap.
    pub fn max_concurrent(mut self, v: i64) -> Self {
        self.max_concurrent = Some(v);
        self
    }

    /// Sets the enforcement scope.
    pub fn scope(mut self, s: Scope) -> Self {
        self.scope = s;
        self
    }

    /// Tunes the number of buckets per window.
    pub fn buckets_per_window(mut self, n: i64) -> Self {
        self.buckets_per_window = n;
        self
    }

    /// Switches the rule to the exact per-event log.
    pub fn precise(mut self, p: bool) -> Self {
        self.precise = p;
        self
    }

    /// Sets a human-readable label.
    pub fn name(mut self, name: &str) -> Self {
        self.name = Some(name.to_string());
        self
    }

    /// The bucket width in seconds: 0 selects the exact per-event log, otherwise
    /// `window_seconds / buckets_per_window`.
    pub fn bucket_seconds(&self) -> f64 {
        if self.precise {
            return 0.0;
        }
        let buckets = self.buckets_per_window.max(1);
        self.window_seconds / buckets as f64
    }

    /// The rule's name, or a derived `model:scope:Ns` label.
    pub fn label(&self) -> String {
        match &self.name {
            Some(n) => n.clone(),
            None => format!(
                "{}:{}:{}s",
                self.model, self.scope, self.window_seconds as i64
            ),
        }
    }

    /// Whether the rule caps tokens, requests, or cost (vs concurrency only).
    pub fn has_usage_limit(&self) -> bool {
        self.max_tokens.is_some() || self.max_requests.is_some() || self.max_cost.is_some()
    }
}

/// Parses a rules document (see `limits.sample.json`) into `LimitRule`s.
pub fn rules_from_json(text: &str) -> Result<Vec<LimitRule>, String> {
    let doc: serde_json::Value = serde_json::from_str(text).map_err(|e| e.to_string())?;
    let arr = doc
        .get("rules")
        .and_then(|v| v.as_array())
        .ok_or_else(|| "document has no \"rules\" array".to_string())?;

    let mut out = Vec::with_capacity(arr.len());
    for row in arr {
        let model = row
            .get("model")
            .and_then(|v| v.as_str())
            .ok_or_else(|| "rule is missing \"model\"".to_string())?;
        let window_seconds = row
            .get("window_seconds")
            .and_then(|v| v.as_f64())
            .unwrap_or(60.0);

        let mut rule = LimitRule::new(model, window_seconds);
        rule.max_tokens = row.get("max_tokens").and_then(|v| v.as_f64());
        rule.max_requests = row.get("max_requests").and_then(|v| v.as_f64());
        rule.max_cost = row.get("max_cost").and_then(|v| v.as_f64());
        rule.max_concurrent = row.get("max_concurrent").and_then(|v| v.as_i64());
        if let Some(scope) = row.get("scope").and_then(|v| v.as_str()) {
            rule.scope = scope.parse::<Scope>()?;
        }
        if let Some(buckets) = row.get("buckets_per_window").and_then(|v| v.as_i64()) {
            rule.buckets_per_window = buckets;
        }
        rule.precise = row
            .get("precise")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        rule.name = row.get("name").and_then(|v| v.as_str()).map(str::to_string);
        out.push(rule);
    }
    Ok(out)
}
